.github/workflows/docker.yaml

```diff
@@ -18,174 +19,106 @@ env:
+        name: Build and push (non PR)
+        uses: docker/bake-action@v5.5.0
+        env:
+          BLUE_ROS_DISTRO: ${{ matrix.ROS_DISTRO }}
+          BLUE_GITHUB_REPO: ${{ steps.lower-repo.outputs.repository }}
         with:
-          context: .
-          file: .docker/Dockerfile
-          build-args: ROS_DISTRO=${{ matrix.ROS_DISTRO }}
-          target: ${{ github.job }}
-          tags: ${{ steps.meta.outputs.tags }}
-          labels: ${{ steps.meta.outputs.labels }}
+          workdir: .docker
+          files: |
+            ./docker-bake.hcl
+            ${{ steps.meta.outputs.bake-file }}
+          targets: |
+            ${{ matrix.stage }}
           push: ${{ env.PUSH }}
-
-  desktop-nvidia:
-    strategy:
-      fail-fast: false
-      matrix:
-        ROS_DISTRO: [rolling]
-    runs-on: ubuntu-latest
-    permissions:
-      packages: write
-      contents: read
-    steps:
-      - name: Checkout repository
-        uses: actions/checkout@v4
-
-      - name: Log into registry
-        if: env.PUSH == 'true'
-        uses: docker/login-action@v3.3.0
+          set: |
+            *.cache-from=type=gha,scope=ci
+            *.cache-from=type=gha,scope=robot
+            *.cache-from=type=gha,scope=desktop
+            *.cache-from=type=gha,scope=desktop-nvidia
+            ${{ matrix.stage }}.cache-to=type=gha,mode=max,scope=${{ matrix.stage }}
+
+      # Pull request builds are not cached; and only built for AMD64
+      - if: github.event_name == 'pull_request'
+        name: Build and push (PR)
+        uses: docker/bake-action@v5.5.0
+        env:
+          BLUE_ROS_DISTRO: ${{ matrix.ROS_DISTRO }}
+          BLUE_GITHUB_REPO: ${{ steps.lower-repo.outputs.repository }}
         with:
-          registry: ghcr.io
-          username: ${{ github.actor }}
-          password: ${{ secrets.GITHUB_TOKEN }}
-
-      - name: Extract Docker metadata
-        if: env.PUSH == 'true'
-        id: meta
-        uses: docker/metadata-action@v5.5.1
-        with:
-          images: ghcr.io/${{ github.repository }}
-          tags: |
-            type=raw,value=${{ matrix.ROS_DISTRO }}-${{ github.job }}
-
-      - name: Build and push Docker image
-        uses: docker/build-push-action@v6.7.0
-        with:
-          context: .
-          file: .docker/Dockerfile
-          build-args: ROS_DISTRO=${{ matrix.ROS_DISTRO }}
-          target: ${{ github.job }}
-          tags: ${{ steps.meta.outputs.tags }}
-          labels: ${{ steps.meta.outputs.labels }}
-          push: ${{ env.PUSH }}
+          workdir: .docker
+          files: |
+            ./docker-bake.hcl
+            ${{ steps.meta.outputs.bake-file }}
+          targets: |
+            ${{ matrix.stage }}
+          set: |
+            *.platform=linux/amd64
+            *.cache-from=type=gha,scope=ci
+            *.cache-from=type=gha,scope=robot
+            *.cache-from=type=gha,scope=desktop
+            *.cache-from=type=gha,scope=desktop-nvidia
```

Comment:
The docs [indicate](https://docs.docker.com/build/cache/backends/) that we need to enable the containerd image store to support gha caching. I accidentally tested the gha caching in my `ros2-template` repo, and received an [error](https://github.com/Robotic-Decision-Making-Lab/ros2-template/actions/runs/10408643095/job/28826497474), indicating it needs to be enabled. [This](https://github.com/docker/setup-buildx-action/issues/257#issuecomment-1722284952) comment from [`setup-buildx-action!257`](https://github.com/docker/setup-buildx-action/issues/257) describes the process to enable it. There may be a better way to use Docker 24 features now though. That issue is about a year old.

Reply:
Hm.  The GHA cache is limited to 10GB for the full repository.  Given the current `{distro}-desktop` image is over 10G for each distro, won't get much mileage from the GHA cache.

Back to registry caching, I think?

Reply:
I believe so, yeah

Reply:
Switched back to using the package registry for cache.   It uses an separate image `cache-$ROS_DISTRO` which has the `mode=max` flag.   I _think_ this has no effect in the current Dockerfile, because it's linear.  However, when there's a build stage (as in #241), it includes the build stage layers in the cache even though they aren't specifically part of the final image(s).

Since it's linear, I also think we can use one cache image.  As it makes it's way through the sequence, each new image is additive on the previous layers.

Reply:
Were you able to do any testing with the cache? Not a problem if not. In the future, if you are implementing a CI feature, you can make a feature branch on the `blue` repo and trigger the pipeline from that branch to test changes are only run on push to `main`. There are also local testing projects like [`act`](https://github.com/nektos/act), but I haven't had much success setting those up.

Reply:
I did, by setting the `cache-to=` line in the PR step (then unsetting it).

I think manual job triggers are a different kind of `event_name` (just checked, it's `workflow_dispatch`), might need to adjust the conditionals to get the right behavior.

Reply:
What I've done in the past to make testing easier is add my feature branch to the `push` job trigger `branches` list, e.g.,
```yaml
  push:
    branches:
      - my-feature-branch
```
I don't believe you can use the the `workflow_dispatch` or `push` triggers from a fork though, which is why I commented about creating a feature branch on the `blue` repo.
![image](https://github.com/user-attachments/assets/6a38f788-39c2-49b8-af97-3a9f4f8cac7b)